Implement `Smt` struct (replacement to `TieredSmt`)

plafer opened the pull request:
Builds on #245

bobbinth commented on src/merkle/smt/full/tests.rs:
```diff
@@ -0,0 +1,278 @@
+#[test]
+fn test_smt_insert_at_same_key_2() {
+    let key_already_present: RpoDigest = {
+        let raw = 0b_01101001_01101100_00011111_11111111_10010110_10010011_11100000_00000000_u64;
```
nit: I think this is the same value as the one on line 75 below (it took me some time to understand this). I'm not sure there is a big benefit to using binary representation here and it actually makes this a bit more difficult to follow. I'd probably use just decimal values for this.

plafer replied:
right - cleaned it up

bobbinth reviewed (approved):
Looks good! Thank you! I left some minor comments inline. After these are addressed, we are good to merge.

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+    inner_nodes: BTreeMap<NodeIndex, InnerNode>,
+}
+
+impl Smt {
```
done

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+
+    /// Inserts a value at the specified key, returning the previous value associated with that key.
+    /// Recall that by definition, any key that hasn't been updated is associated with
+    /// [`EMPTY_WORD`].
```
done

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Returns a new [NewSmt].
+    ///
+    /// All leaves in the returned tree are set to [ZERO; 4].
+    pub fn new() -> Self {
```
done

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Returns an error if:
+    /// - The number of entries exceeds 2^63 entries.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
```
done

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+    /// Remove key-value pair into the leaf stored at key; return the previous value associated with
+    /// `key`, if any. We also return an `is_empty` flag which indicates whether the leaf became
+    /// empty, and must be removed from the data structure it is contained in.
+    fn remove(&mut self, key: RpoDigest) -> (Option<Word>, bool) {
```
fixed docstring

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+
+    /// Insert key-value pair into the leaf; return the previous value associated with `key`, if
+    /// any.
+    fn insert(&mut self, key: RpoDigest, value: Word) -> Option<Word> {
```
done

plafer commented on src/merkle/smt/full/mod.rs:
```diff
@@ -0,0 +1,366 @@
+    }
+
+    /// Compute the hash of the leaf
+    pub fn hash(&self) -> RpoDigest {
```
done

plafer commented:
@bobbinth addressed PR comments, and fixed `SmtLeaf` as discussed in [a880c88](https://github.com/0xPolygonMiden/crypto/pull/254/commits/a880c880ab4f9f22507de24536b99f0700f61ebb)

bobbinth reviewed (approved):
All looks good! Thank you!
